posthog/models/sessions/sql.py

```diff
@@ -0,0 +1,131 @@
+    _timestamp SimpleAggregateFunction(max, DateTime)
+) ENGINE = {engine}
+"""
+
+SESSIONS_DATA_TABLE_ENGINE = lambda: AggregatingMergeTree(TABLE_BASE_NAME, replication_scheme=ReplicationScheme.SHARDED)
+
+SESSIONS_TABLE_SQL = lambda: (
+    SESSIONS_TABLE_BASE_SQL
+    + """
+    PARTITION BY toYYYYMM(min_first_timestamp)
+    -- order by is used by the aggregating merge tree engine to
+    -- identify candidates to merge, e.g. toDate(min_first_timestamp)
+    -- would mean we would have one row per day per session_id
+    -- if CH could completely merge to match the order by
+    -- it is also used to organise data to make queries faster
+    -- we want the fewest rows possible but also the fastest queries
+    -- since we query by date and not by time
+    -- and order by must be in order of increasing cardinality
+    -- so we order by date first, then team_id, then session_id
+    -- hopefully, this is a good balance between the two
+    ORDER BY (toDate(min_first_timestamp), team_id, session_id)
+SETTINGS index_granularity=512
+"""
+).format(
+    table_name=SESSIONS_DATA_TABLE(),
+    cluster=settings.CLICKHOUSE_CLUSTER,
+    engine=SESSIONS_DATA_TABLE_ENGINE(),
+)
+
+SESSIONS_TABLE_MV_SQL = (
+    lambda: """
+CREATE MATERIALIZED VIEW IF NOT EXISTS {table_name} ON CLUSTER '{cluster}'
+TO {database}.{target_table}
+AS SELECT
+$session_id as session_id,
+team_id,
+any(distinct_id) as distinct_id,
+min(timestamp) AS min_first_timestamp,
+max(timestamp) AS max_last_timestamp,
+-- TRICKY: ClickHouse will pick a relatively random "entry"
+-- when it collapses the aggregating merge tree
+-- unless we teach it what we want...
+-- argMin ignores null values
+-- so this will get the first non-null value of "entry"
+-- for each group of session_id and team_id
+-- by min of first_timestamp in the batch
+-- this is an aggregate function, not a simple aggregate function
+-- so we have to write to argMinState, and query with argMinMerge
+
+groupUniqArray(JSONExtractString(properties, '$current_url')) AS urls,
+argMinState(JSONExtractString(properties, '$current_url'), timestamp) as entry_url,
+argMaxState(JSONExtractString(properties, '$current_url'), timestamp) as exit_url,
+argMinState(JSONExtractString(properties, '$initial_utm_source'), timestamp) as initial_utm_source,
+argMinState(JSONExtractString(properties, '$initial_utm_campaign'), timestamp) as initial_utm_campaign,
+argMinState(JSONExtractString(properties, '$initial_utm_medium'), timestamp) as initial_utm_medium,
+argMinState(JSONExtractString(properties, '$initial_utm_term'), timestamp) as initial_utm_term,
+argMinState(JSONExtractString(properties, '$initial_utm_content'), timestamp) as initial_utm_content,
+argMinState(JSONExtractString(properties, '$initial_referring_domain'), timestamp) as initial_referring_domain,
+count(*) as event_count,
+sumIf(1, event='$pageview') as pageview_count
+
+FROM {database}.events
+group by session_id, team_id
```

Comment:
I think you' want the where timestamp > x here so that you know backfills always have a known roof to run up to

i think it doesn't massively matter because the table will dedupe anything that is identical across its sort key but there's not point doing work for no reason

Reply:
re-reading that doc https://den-crane.github.io/Everything_you_should_know_about_materialized_views_commented.pdf

you create the MV with a date in the future so that you know all data from before that date hasn't be read into the materialized view and you can backfill anything before that date safely

Reply:
Yeah I was just reading that. It's more about "this is a good idea to make it clear for the future" rather than actually affecting the way the view works.

Reply:
> rather than actually affecting the way the view works.

well... it'll stop data from before that date entering the view, no?

Reply:
Right yes indeed but I meant more that it's not like omitting it would lead to a backfill or something

Reply:
Yep, it's just so you know when you backfill you can backfill whole days without duplication 👍